### PR TITLE
Implement several new metrics for speech recognition

### PR DESCRIPTION
**Pull request by asumagic**
## What does this PR do?

**VERY WIP.**

The goal is to introduce a number of metrics:

- Tools to help define "synonym" dictionaries for words that should be considered identical for the WER (by allowing to define an equality function during the WER calculation)
- Various metrics introduced and/or suggested by [Qualitative Evaluation of Language Model Rescoring in Automatic Speech
Recognition](https://www.isca-archive.org/interspeech_2022/roux22_interspeech.pdf):
  - Part-of-speech Error Rate (POSER)
  - Lemma Error Rate (LER)
  - Embedding Error Rate (EmbER)
  - BERTScore
  - Sentence Semantic Distance (SemDist)

<details>
  <summary><b>Before submitting</b></summary>

- [ ] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing [tests](https://github.com/speechbrain/speechbrain/tree/develop/tests) pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [ ] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [ ] Review the self-review checklist to ensure the code is ready for review

</details>

### CODE REVIEW

**Comment by asumagic** [comment]
Tests seem to fail because `speechbrain/SSL_Quantization` on HF returns 401, should stuff requiring HF even be part of the doctests?

**Comment by asumagic** [comment]
Added a link to a tutorial I just finished completing in the main post.

**Comment by asumagic** [comment]
Force-pushed rebase to match develop CI.

**Comment by asumagic** [comment]
Ready for review, although the CI is disagreeing for reasons that are beyond me, considering `develop` passes CI perfectly fine.

File: .github/workflows/pythonapp.yml [hunk 1]
```diff
@@ -44,6 +44,8 @@ jobs:
                   pip install k2==1.24.4.dev20231220+cpu.torch2.1.2 -f https://k2-fsa.github.io/k2/cpu.html
                   pip install protobuf
                   pip install kaldilm==1.15
+                  pip install spacy
+                  pip install flair
```
Comment (Adel-Moumen) [inline .github/workflows/pythonapp.yml:48]: oof maybe we should freeze them

File: docs/docs-requirements.txt [hunk 2]
```diff
@@ -1,12 +1,14 @@
 better-apidoc>=0.3.1
 ctc-segmentation>=1.7.0
 fairseq
+flair
```
Comment (Adel-Moumen) [inline docs/docs-requirements.txt:4]: same as before

File: docs/docs-requirements.txt [hunk 3]
```diff
@@ -1,12 +1,14 @@
...
 https://github.com/kpu/kenlm/archive/master.zip
 numba>=0.54.1
 pyctcdecode
 recommonmark>=0.7.1
 scikit-learn
 six
+spacy
```
Comment (Adel-Moumen) [inline docs/docs-requirements.txt:11]: same as before

File: speechbrain/lobes/models/flair/embeddings.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        filename : str, optional
+            The filename of the model. The default is the usual filename for
+            this kind of model.
+        *args
+            Extra positional arguments to pass to the flair class constructor
+        **kwargs
+            Extra keyword arguments to pass to the flair class constructor
+        """
+
+        target = save_path + "/flair-emb--" + source.replace("/", "--") + "/"
```
Comment (Adel-Moumen) [inline speechbrain/lobes/models/flair/embeddings.py:61]: i suppose that the target template is what flair is using right? Maybe we should leave it as an optional argument? I don't know much about flair but maybe this could make sense. Additionally, I would say you should add a small comment explaining what you are doing

File: speechbrain/lobes/models/flair/sequencetagger.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            The location of the model (a directory or HF repo, for instance).
+        save_path : str, optional
+            The saving location for the model (i.e. the root for the download or
+            symlink location).
+        filename : str, optional
+            The filename of the model. The default is the usual filename for
+            this kind of model.
+        """
+
+        target = save_path + "/flair--" + source.replace("/", "--") + "/"
```
Comment (Adel-Moumen) [inline speechbrain/lobes/models/flair/sequencetagger.py:50]: same remark as before

File: speechbrain/lobes/models/spacy/nlp.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    list of list of str
+        For each sentence, the sequence of extracted lemmas as `str`s."""
+    return [[tok.lemma_ for tok in doc] for doc in docs]
+
+
+class SpacyPipeline:
+    """Modelizes a spaCy pipeline."""
+
+    def __init__(self, nlp):
+        self.nlp = nlp
```
Comment (Adel-Moumen) [inline speechbrain/lobes/models/spacy/nlp.py:54]: missing docstring for input args. 

Wondering if we should add an example as well of how to use it.

File: speechbrain/utils/bertscore.py [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+    evaluation with many different models.
+
+    .. warning::
+        Out of the box, this implementation may not strictly match the results
+        of the reference implementation. Please read the argument documentation
+        to understand the differences.
+
+    Arguments
+    ---------
+    model : transformers.AutoModelForTextEncoding
```
Comment (Adel-Moumen) [inline speechbrain/utils/bertscore.py:43]: rename model to lm

File: speechbrain/utils/bertscore.py [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+        tokenizer,
+        batch_size: int = 64,
+        use_idf: bool = True,
+        sentence_level_averaging: bool = True,
+        num_layers: Optional[int] = None,
+        allow_matching_special_tokens: bool = False,
+    ):
+        self.clear()
+        self.lm = lm
+        self.tokenizer = tokenizer
```
Comment (Adel-Moumen) [inline speechbrain/utils/bertscore.py:90]: okay what about using our HFTransformer interface for loading the LM/tokenizer automatically?

File: speechbrain/utils/bertscore.py [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+
+        with torch.no_grad():
+            self._update_summary()
+
+        if field is not None:
+            return self.summary[field]
+
+        return self.summary
+
+    def _update_summary(self):
```
Comment (Adel-Moumen) [inline speechbrain/utils/bertscore.py:155]: could you please add a small docstring ?

File: speechbrain/utils/bertscore.py [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+
+            # mask off weights
+            recall_weights[~ref_mask] = 0.0
+            precision_weights[~hyp_mask] = 0.0
+
+            batch_recall = recall_values * recall_weights
+            batch_precision = precision_values * precision_weights
+
+            for i, utt_id in enumerate(ids):
+                # TODO: optionally provide a token->token map
```
Comment (Adel-Moumen) [inline speechbrain/utils/bertscore.py:221]: ? :p

File: speechbrain/utils/bertscore.py [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+    from the reference corpus are weighted with `+1` smoothing.
+
+    Arguments
+    ---------
+    tokenizer
+        HuggingFace tokenizer for the BERT model.
+    corpus : Iterable[str], optional
+        Iterable corpus to compute the IDF from. Each iterated value is
+        considered a document in the corpus in the IDF calculation.
+        If omitted, no IDF weighting is done.
```
Comment (Adel-Moumen) [inline speechbrain/utils/bertscore.py:347]: could you please add a Return docstring?

File: speechbrain/utils/dictionaries.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+    @staticmethod
+    def from_json_file(file) -> "SynonymDictionary":
+        """Parses an opened file as JSON, where the top level structure is a
+        list of sets of synonyms (i.e. words that are all synonyms with each
+        other), e.g. `[ ["hello", "hi"], ["say", "speak", "talk"] ]`.
+
+        Arguments
+        ---------
+        file
+            File object that supports reading (e.g. an `open`ed file)
```
Comment (Adel-Moumen) [inline speechbrain/utils/dictionaries.py:34]: return docstring

File: speechbrain/utils/dictionaries.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+
+    @staticmethod
+    def from_json_path(path) -> "SynonymDictionary":
+        """Opens a file and parses it as JSON, with otherwise the same semantics
+        as :meth:`~SynonymDictionary.from_json_file`, which uses an opened file.
+
+        Arguments
+        ---------
+        path : str
+            Path to the JSON file
```
Comment (Adel-Moumen) [inline speechbrain/utils/dictionaries.py:58]: return

File: speechbrain/utils/dictionaries.py [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+    def __call__(self, a: str, b: str) -> bool:
+        """Check for the equality or synonym equality of two words.
+
+        Arguments
+        ---------
+        a : str
+            First word to compare. May be outside of the known dictionary.
+        b : str
+            Second word to compare. May be outside of the known dictionary.
+            The order of arguments does not matter."""
```
Comment (Adel-Moumen) [inline speechbrain/utils/dictionaries.py:85]: return as well please

File: speechbrain/utils/distances.py [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    # here the matrix multiply effectively results, for [..., x, y], in the dot
+    # product of the normalized `a[..., x, :]` and `b[..., y, :]` vectors, thus
+    # giving us the proper cosine similarity.
+    # multiplication shape: a[..., X, 1] @ b[..., 1, Y]
+    return a_normalized @ b_normalized.transpose(-1, -2)
+
+    # NOTE: needs a test written:
+    # - case where a dim is zero
+    # - case of 2-dim
+    # - case of more dims
```
Comment (Adel-Moumen) [inline speechbrain/utils/distances.py:55]: blbl?

File: speechbrain/utils/metric_stats.py [hunk 16]
```diff
@@ -358,10 +374,235 @@ def write_stats(self, filestream):
...
+    def append(self, *args, **kwargs):
+        """Append function, which should **NOT** be used for the weighted error
+        rate stats. Please append to the specified `base_stats` instead.
+
+        Arguments
+        ---------
+        *args
+            Ignored.
+        **kwargs
+            Ignored.
```
Comment (Adel-Moumen) [inline speechbrain/utils/metric_stats.py:417]: not clear to me :p

File: speechbrain/utils/metric_stats.py [hunk 17]
```diff
@@ -358,10 +374,235 @@ def write_stats(self, filestream):
...
+            "You should only append to the base ErrorRateStats."
+        )
+
+    def summarize(self, field=None):
+        """Returns a dict containing some detailed WER statistics after weighting
+        every edit with a weight determined by `cost_function` (returning `0.0`
+        for no error, `1.0` for the default error behavior, and anything in
+        between).
+
+        See :meth:`~ErrorRateStats.summarize`."""
```
Comment (Adel-Moumen) [inline speechbrain/utils/metric_stats.py:431]: I would say we need to imrpove the docstring to better reflect input / output variables

File: speechbrain/utils/metric_stats.py [hunk 18]
```diff
@@ -358,10 +374,235 @@ def write_stats(self, filestream):
...
+        returned by `summarize`.
+        See :meth:`~ErrorRateStats.write_stats`.
+        """
+        if not self.summary:
+            self.summarize()
+
+        print(f"Weighted WER metrics ({self.weight_name}):", file=filestream)
+
+        for k, v in self.summary.items():
+            print(f"{k}: {v}", file=filestream)
```
Comment (Adel-Moumen) [inline speechbrain/utils/metric_stats.py:519]: logger vs print ?

File: speechbrain/utils/metric_stats.py [hunk 19]
```diff
@@ -358,10 +374,235 @@ def write_stats(self, filestream):
...
+        high_similarity_weight: float,
+        threshold: float,
+    ):
+        self.embedding_function = embedding_function
+        self.low_similarity_weight = low_similarity_weight
+        self.high_similarity_weight = high_similarity_weight
+        self.threshold = threshold
+
+    def __call__(
+        self, edit_symbol: str, a: Optional[str], b: Optional[str]
```
Comment (Adel-Moumen) [inline speechbrain/utils/metric_stats.py:570]: docstring please

File: speechbrain/utils/semdist.py [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+    model : transformers.AutoModelForTextEncoding
+        Transformers text encoder. May live on a non-CPU device.
+    tokenizer : transformers.AutoTokenizer
+        Transformers tokenizer.
+    embed_function : Callable[[List[str]], torch.Tensor]
+        Given a list of sentences, return their summarized embedding using the
+        method of your choice (e.g. mean pooling)
+    scale : float, optional
+        The `α` scale applied to the cosine similarity result for clarity. The
+        default is `1000`, in order to match the authors' recommendation.
```
Comment (Adel-Moumen) [inline speechbrain/utils/semdist.py:30]: the docstring is not updated with respect to the __init__ args.

File: speechbrain/utils/semdist.py [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+
+        with torch.no_grad():
+            self._update_summary()
+
+        if field is not None:
+            return self.summary[field]
+
+        return self.summary
+
+    def _update_summary(self):
```
Comment (Adel-Moumen) [inline speechbrain/utils/semdist.py:82]: one line docstring pls

File: speechbrain/utils/semdist.py [hunk 22]
```diff
@@ -0,0 +1,163 @@
...
+            the model's predictions in tokenizable format
+        targets: list
+            the ground truths in tokenizable format
+        """
+        self.ids.extend(ids)
+        self.predictions.extend(predict)
+        self.targets.extend(target)
+
+    def summarize(self, field=None):
+        """Summarize the SemDist metric scores. Performs the actual embedding
```
Comment (Adel-Moumen) [inline speechbrain/utils/semdist.py:71]: input/output docstring pls

File: speechbrain/utils/semdist.py [hunk 23]
```diff
@@ -0,0 +1,163 @@
...
+class SemDistStats(BaseSemDistStats):
+    """Computes the SemDist metric with a provided HuggingFace Transformers
+    tokenizer and LM.
+
+    Arguments
+    ---------
+    model : transformers.AutoModelForTextEncoding
+        Transformers text encoder. May live on a non-CPU device.
+    tokenizer : transformers.AutoTokenizer
+        Transformers tokenizer.
```
Comment (Adel-Moumen) [inline speechbrain/utils/semdist.py:118]: these parameters are not the one used in the __init__

File: speechbrain/utils/semdist.py [hunk 24]
```diff
@@ -0,0 +1,163 @@
...
+        self,
+        lm,
+        tokenizer,
+        method: Literal["meanpool", "cls"] = "meanpool",
+        *args,
+        **kwargs,
+    ):
+        super().__init__(embed_function=self._embed, *args, **kwargs)
+        self.lm = lm
+        self.tokenizer = tokenizer
```
Comment (Adel-Moumen) [inline speechbrain/utils/semdist.py:140]: same comment as before, maybe we could use HFTransformer instead.

File: speechbrain/utils/semdist.py [hunk 25]
```diff
@@ -0,0 +1,163 @@
...
+        self.method = method
+
+    def _embed(self, sentences: List[str]) -> torch.Tensor:
```
Comment (Adel-Moumen) [inline speechbrain/utils/semdist.py:143]: docstring :p

**Review by Adel-Moumen — COMMENTED** [review]
Hi, 

First, this is a really good work. I left some small comments that are mostly related to docstring. Also, I was wondering if it would make sense to use our HF interface to load the LM/tokenizer directly meaning that the user would only have to pass in input a string reflecting the path of the HF model. 

Also, could you please add a recipe that shows how to use the new metrics? e.g. take LibriSpeech transformer for instance. 

Also, could you please make sure to update this PR w.r.t speechbrain/develop? I suppose that you will have to fix some issues due to the new merged PR on linters. 

Thanks again this is good work.

**Review by asumagic — COMMENTED** [review]
Updated the tutorial with the new `TextEncoder` HF interface. I believe this should be ready to review again.

**Comment by asumagic** [comment]
That said, I still don't understand why CI is failing.

**Comment by Adel-Moumen** [comment]
> That said, I still don't understand why CI is failing.

The error is due to https://huggingface.co/speechbrain/SSL_Quantization being private. Could you please add doctest skip in the example so that It skip running the example?

File: speechbrain/lobes/models/flair/embeddings.py [hunk 26]
```diff
@@ -0,0 +1,128 @@
...
+        return torch.stack(sentence_embs)
+
+    def embed_word(self, word: str) -> torch.Tensor:
+        """Embeds a single word.
+
+        Arguments
+        ---------
+        word : str
+            Word to embed. Out-of-vocabulary handling depends on the underlying
+            embedding class.
```
Comment (Adel-Moumen) [inline speechbrain/lobes/models/flair/embeddings.py:125]: funny that the ci is not complaining about this docstring which does not have "returns" docstring. Do you have fetch the latest version of spechbrain?

File: speechbrain/lobes/models/huggingface_transformers/textencoder.py [hunk 27]
```diff
@@ -0,0 +1,125 @@
...
+
+from speechbrain.lobes.models.huggingface_transformers.huggingface import (
+    HFTransformersInterface,
+)
+
+from typing import Optional
+
+logger = logging.getLogger(__name__)
+
+os.environ["TOKENIZERS_PARALLELISM"] = "false"
```
Comment (Adel-Moumen) [inline speechbrain/lobes/models/huggingface_transformers/textencoder.py:23]: What is that?

**Review by Adel-Moumen — APPROVED** [review]
The code looks clean! I am only wondering why the CI is not asking you to complte the "returns" section of some docstring since it should be required but other than it looks good to me.

**Comment by Adel-Moumen** [comment]
As for the tutorial, could you please try to stick with the shape of SpeechBrain colabs? (same header etc)

File: speechbrain/lobes/models/huggingface_transformers/textencoder.py (hunk 27 above)
Comment (asumagic) [reply speechbrain/lobes/models/huggingface_transformers/textencoder.py:23]: Good point, this is an artifact from using another file in the module as a reference... It seems to be present for several others, and I don't know what the actual rationale for it was.
